jtuples/src/main/java/org/jtuples/Functional.java

```diff
@@ -0,0 +1,210 @@
+    public static <A, B, C> Function<A, Pair<B, C>> split(
+            Function<A, B> f, Function<? super A, C> g) {
+        return (A a) -> new Pair<>(f.apply(a), g.apply(a));
+    }
+
+    /**
+     * Returns the split of the two given functions.
+     * Given two functions with parameters of the same type, the split of
+     * such functions is a single function, with the same parameters, that
+     * returns a pair whose components are the application of the original
+     * functions to the arguments.
+     * Note that this, in a way, is equivalent to the product of the given
+     * functions, although simplified.
+     * @param <A> the type of the first parameter
+     * @param <B> the type of the second parameter
+     * @param <C> the return type of the first function
+     * @param <D> the return type of the second function
+     * @param f the first function of the split
+     * @param g the second function of the split
+     * @return the split of the given functions
+     */
+    public static <A, B, C, D> BiFunction<A, B, Pair<C, D>> split(
+            BiFunction<A, B, C> f, BiFunction<? super A, ? super B, D> g) {
+        return (A a, B b) -> new Pair<>(f.apply(a, b), g.apply(a, b));
+    }
+
+    /**
+     * Returns the split of the two given functions.
+     * Given two functions with parameters of the same type, the split of
+     * such functions is a single function, with the same parameters, that
+     * returns a pair whose components are the application of the original
+     * functions to the arguments.
+     * Note that this is equivalent to the product of suppliers.
+     * @param <A> the return type of the first function
+     * @param <B> the return type of the second function
+     * @param f the first function of the split
+     * @param g the second function of the split
+     * @return the split of the given functions
+     */
+    public static <A, B> Supplier<Pair<A, B>> split(
+            Supplier<A> f, Supplier<B> g) {
+        return product(f, g);
+    }
+
+    /**
+     * Returns the product of the two given functions.
+     * Given two functions, the product of such functions is a single function,
+     * with the union of the parameters, as a Pair, that
+     * returns a pair whose components are the application of the original
+     * functions to the respective arguments.
+     * @param <A> the type of the parameter of the first function
+     * @param <B> the return type of the first function
+     * @param <C> the type of the parameter of the second function
+     * @param <D> the return type of the second function
+     * @param f the first function of the product
+     * @param g the second function of the product
+     * @return the product of the given functions
+     */
+    public static <A, B, C, D> Function<Pair<A, C>, Pair<B, D>> product(
+            Function<A, B> f, Function<C, D> g) {
+        return (Pair<A, C> p) ->
+            new Pair<>(f.apply(p.first()), g.apply(p.second()));
```

Comment:
It *can* be done with split, but I'm failing to see how it would improve the readability, or even the performance. You'd need to compose functions, and add a function to Pair, or write a lambda, that, given a pair, extracts the first/second.